### PR TITLE
New game settings

### PR DESCRIPTION
**Pull request by Oliveriver**
**Brief Description of What This PR Does**

Adds a new game settings screen to the main menu, with basic and advanced views, and a multitude of options which affect in-game content.

See: https://forum.revolutionarygamesstudio.com/t/implementing-new-game-settings/884

**Related Issues**

Closes https://github.com/Revolutionary-Games/Thrive/issues/873
Closes https://github.com/Revolutionary-Games/Thrive/issues/575

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Comment by Oliveriver** [comment]
Note that this PR is now blocked by https://github.com/Revolutionary-Games/Thrive/pull/3278 and https://github.com/Revolutionary-Games/Thrive/pull/3235.

**Comment by Kasterisk** [comment]
I haven't checked in this PR but this discussion https://github.com/Revolutionary-Games/Thrive/pull/3385#discussion_r890938068 relating to the new game button tooltip could be applied here if not yet.

**Comment by Oliveriver** [comment]
> I haven't checked in this PR but this discussion [#3385 (comment)](https://github.com/Revolutionary-Games/Thrive/pull/3385#discussion_r890938068) relating to the new game button tooltip could be applied here if not yet.

Ah, I haven’t added tooltips yet. I’ll keep that in mind when I get to it.

**Comment by Oliveriver** [comment]
I've decided that, with https://github.com/Revolutionary-Games/Thrive/pull/3235 done, this PR can go ahead without https://github.com/Revolutionary-Games/Thrive/pull/3278. I've therefore made the procedural/classic map selection invisible, with the intention of making it visible as part of the procedural patch map PR (same for Easter eggs).

So other than the spawn system weirdness I mentioned on Discord, I believe this PR is ready for review and testing. Testing should cover both checking the settings change what they say they change and deciding what values they should take for difficulty presets.

I doubt we'll get much effective feedback on the latter until we open this up to the community, seeing as we as devs have a skewed experience of the game's difficulty. So I'd like to request that we get this into a devbuild sooner rather than later (this is why I've changed course to not depend on the procedural patch map).

File: src/general/NewGameSettings.cs
Comment (Kasterisk) [inline src/general/NewGameSettings.cs:348]: The usual non-written way we do it is to only call the play button press sound from the callbacks, so that the actual logic containing methods doesn't have to know whether to play the sound or not.

File: src/general/NewGameSettings.cs
Comment (Kasterisk) [inline src/general/NewGameSettings.cs:398]: This doesn't seem very good, there's now duplicate main menu instances and since this one isn't added to the scene tree, it will be orphaned, potentially causing a memory leak. Instead, I think it's better to just move the logic in `MainMenu.NewGameSetupDone` to here. It's pertaining to the new game settings anyways.

File: src/general/NewGameSettings.cs
Comment (Kasterisk) [inline src/general/NewGameSettings.cs:568]: Wouldn't it be far more efficient to cast these indexes directly to enum?
```csharp
settings.Origin = (WorldGenerationSettings.LifeOrigin)index;
```

**Review by Kasterisk — DISMISSED** [review]

**Comment by Kasterisk** [comment]
Really like how the UI works. I left some reviews for things I think could be improved codewise.

File: src/general/WorldGenerationSettings.cs
Comment (Oliveriver) [reply src/general/WorldGenerationSettings.cs:23]: I thought it might be cleaner to keep these constants with the rest of the world generation settings, since they're exclusively applicable here. But if the consensus says otherwise, I'll move them.

File: src/general/NewGameSettings.cs
Comment (Oliveriver) [reply src/general/NewGameSettings.cs:568]: Did not know this was possible. Will definitely change that.

File: src/microbe_stage/PatchMapGenerator.cs
Comment (Oliveriver) [reply src/microbe_stage/PatchMapGenerator.cs:11]: All the methods in this class are static, so this has to be static.

**Review by RacerBG — COMMENTED** [review]
I'm still in the early stage of testing this PR so I will give my final verdict later. So far I'm very impressed with the new UI, it looks professional, nice work!

I would suggest having a tooltip over the random seed button though.

"Advanced view" and "Basic view" could be renamed in to "Basic" and "Advanced" but it's up to you.

I'm not so sure about the text below the multicellular stage because saving and loading should be working quite soon in my opinion.

I had tested the values from "Advanced view" and they seem to work just fine but I would like to spend a few hours on them tomorrow and then give my final review. :smiley:

File: src/general/WorldGenerationSettings.cs
Comment (hhyyrylainen) [reply src/general/WorldGenerationSettings.cs]: I think maybe in this case the huge dump of difficulty option ranges should maybe be in the separate file. At least if they were moved to constants they'd need much more descriptive names, and I imagine (I haven't looked through this PR yet) that only the world generation really should use these. So the design is a bit clearer if other places in the code aren't able to access these and instead need to read the values that the current game has stored.

File: src/microbe_stage/OrganelleDefinition.cs
Comment (hhyyrylainen) [reply src/microbe_stage/OrganelleDefinition.cs]: I think naming this to LAWK would make more sense as it is not a normal word (need to also add to the jetbrains dotsettings file so that it doesn't say this is mispelled).

File: src/microbe_stage/PatchMapGenerator.cs
Comment (hhyyrylainen) [reply src/microbe_stage/PatchMapGenerator.cs:11]: Why are the settings static here? Shouldn't they be a generation run specific?

File: src/general/GameWorld.cs
Comment (hhyyrylainen) [reply src/general/GameWorld.cs:21]: Yeah I agree. Only one of them should have the variable. And well I think I can see the use of storing the world generation settings, but in my mind I was imagining that this object would *only* exist while generating the world. But then again later on in the game we'll also do world generation so I guess storing this somewhere with the current game data is the right design.

File: assets/concept_art/planet_generation.jpeg.import
Comment (hhyyrylainen) [inline assets/concept_art/planet_generation.jpeg.import:1]: For uniformity I think this should be a .jpg file (we do have one .jpeg file in the game, but that's more of an oversight).

File: simulation_parameters/SimulationParameters.cs
Comment (hhyyrylainen) [inline simulation_parameters/SimulationParameters.cs:233]: This ToList seems to unnecessarily create a temporary list here. If you change `var` in used organelle to `IEnumerable<OrganelleDefinition>` I think this code will compile without the ToList call here.

File: simulation_parameters/SimulationParameters.cs
Comment (hhyyrylainen) [inline simulation_parameters/SimulationParameters.cs:255]: Same here

File: src/auto-evo/steps/FindBestMutation.cs
Comment (hhyyrylainen) [inline src/auto-evo/steps/FindBestMutation.cs:23]: I think here the use of WorldGenerationSettings is pretty unclear why it is needed here. I think the design would be clearer if you create some interface to store the difficulty data or the world configuration that is needed here. Something like IWorldConfiguration or IDifficultyOption or something like that, or maybe even passing in the current game here?

File: src/microbe_stage/PatchMapGenerator.cs
Comment (Oliveriver) [reply src/microbe_stage/PatchMapGenerator.cs:11]: As I said, everything in this class is static for some reason, so an instance variable wouldn’t work.

Although now that I think about it, there’s probably no need to store them as a member variable here since they’re passed in every time. I’ll change that to resolve this.

File: src/auto-evo/steps/FindBestMutation.cs
Comment (Oliveriver) [reply src/auto-evo/steps/FindBestMutation.cs:23]: Passing in the current game sounds simplest to me.

File: src/microbe_stage/PatchMapGenerator.cs
Comment (hhyyrylainen) [reply src/microbe_stage/PatchMapGenerator.cs:11]: If the design starts to look like each method needs to pass the current settings all around, I think converting this to a non-static class would start to make a lot of sense so that the settings can be passed just once to the constructor.

File: src/general/WorldGenerationSettings.cs
Comment (hhyyrylainen) [reply src/general/WorldGenerationSettings.cs]: I think they might make sense here (I haven't read the code for this PR yet), another place might be the new game GUI, since these look like to be the intervals the new game options need to be within. That way it would be super clear that only that new game GUI should use this (especially if these could be made private in there), and all other places in the code need to access the actual current game variables.

File: simulation_parameters/SimulationParameters.cs
Comment (Oliveriver) [reply simulation_parameters/SimulationParameters.cs:233]: Attempting to do this is giving me a PossibleMultipleEnumeration error during style checking. Any suggestions or should I keep it as is?

File: simulation_parameters/SimulationParameters.cs
Comment (hhyyrylainen) [inline simulation_parameters/SimulationParameters.cs:234]: Don't do nullability suppression like this. It is only meant for constructor-like resources (like Godot _Ready). In this case ~~I think you can just leave this variable uninitialized, or~~ just assign null (you need to make the variable nullable first `OrganelleDefinition?`). And then after the loop if the variable is null throw an exception as it's invalid that there wasn't any organelles.

File: src/auto-evo/steps/FindBestMutation.cs
Comment (hhyyrylainen) [inline src/auto-evo/steps/FindBestMutation.cs:22]: It seems that the game world is used pretty often so maybe AutoEvoConfiguration should have the world as a mandatory constructor parameter so it'd be available everywhere in auto-evo or maybe at least the world settings should be in there. As only needing world settings will be easier to use for auto-evo prediction and the auto-evo exploring tool, I imagine, when that is done.

File: src/general/MainMenu.cs
Comment (hhyyrylainen) [inline src/general/MainMenu.cs:76]: I'd make this a non-static method to be able to put this later in the class where the other new game methods are, to not have this be the first piece of code someone sees when they try to read the main menu file.

Comment (hhyyrylainen) [reply src/general/MainMenu.cs:76]: Okay, I read now this entire file... I guess it's not worth trying to think of a different design for this as this is needed to be called from the new game menu as there's other stuff to fix in this PR...

File: src/general/Mutations.cs
Comment (hhyyrylainen) [inline src/general/Mutations.cs:168]: Why is the creationRate a double here?

File: src/general/WorldGenerationSettings.cs
Comment (hhyyrylainen) [inline src/general/WorldGenerationSettings.cs:23]:
```suggestion
    // Values for this particular object
```
This is the comment style used in the code, C-style comments are not used.

Comment (Oliveriver) [reply src/general/WorldGenerationSettings.cs:23]: I used this style because it turned up a few times here, so maybe this should be changed too: https://github.com/Revolutionary-Games/Thrive/blob/master/src/general/OptionsMenu.cs

File: src/general/WorldGenerationSettings.cs
Comment (hhyyrylainen) [inline src/general/WorldGenerationSettings.cs:25]: Most of these are pretty self explanatory, but could you add a short summary for each? These are pretty widely affecting things so similarly to the game settings object, I think all of these properties should be documented.

Also I think this field should be named LAWK.

Comment (Oliveriver) [reply src/general/WorldGenerationSettings.cs:25]: Code styling had a problem with the name LAWK so I changed it to this.

File: src/general/WorldGenerationSettings.cs
Comment (hhyyrylainen) [reply src/general/WorldGenerationSettings.cs:23]: Opened a PR: https://github.com/Revolutionary-Games/Thrive/pull/3422

File: src/general/WorldGenerationSettings.cs
Comment (hhyyrylainen) [reply src/general/WorldGenerationSettings.cs:25]: You need to add the LAWK as an abbreviation so that it knows that it is a single word and stops complaining.

Comment (hhyyrylainen) [reply src/general/WorldGenerationSettings.cs:25]: Thrive.sln.DotSettings around line 460

File: src/general/WorldGenerationSettings.cs
Comment (hhyyrylainen) [inline src/general/WorldGenerationSettings.cs:44]: I think it would be a cleaner design if there was a Difficulty class which already had all of the right properties initialized for the difficulty level, that way these constant lookups would not be needed, and those difficulty presets could be loaded from a JSON file for easy tweaking by @buckly90 for balancing.

File: src/auto-evo/steps/FindBestMutation.cs
Comment (hhyyrylainen) [reply src/auto-evo/steps/FindBestMutation.cs:22]: You are right, I was thinking of `SimulationConfiguration` which doesn't seem to be passed as a parameter here so the data can't be added there to easily use here.

**Review by RacerBG — APPROVED** [review]
Alright, part II of my review and my approval of course:

I don't have a strong opinion about the values affected by the difficulty they are looking fine in my opinion but you are free to change them as you like. :) The overall experience is pleasant, my earlier comments were mostly adressed and I can't find any game breaking bugs either.

File: src/auto-evo/steps/IncreaseBiodiversity.cs
Comment (hhyyrylainen) [inline src/auto-evo/steps/IncreaseBiodiversity.cs:26]: I'd suggest naming `settings` `worldSettings` here so that there aren't two very similarly named things, configuration and settings.

File: simulation_parameters/SimulationParameters.cs
Comment (hhyyrylainen) [inline simulation_parameters/SimulationParameters.cs:236]:
```suggestion
        return difficultyPresets.Values.First(p => p.Index == index);
```
I think that's valid LINQ.

File: src/general/DifficultyPreset.cs
Comment (hhyyrylainen) [inline src/general/DifficultyPreset.cs:9]:
```suggestion
///     Values for each difficulty preset, as given in difficulty_presets.json
```
typo

File: src/microbe_stage/MicrobeStage.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeStage.cs:123]: Shouldn't this be json ignore and just have a getter for getting the object from the gameworld? Similarly to how the gameworld is just a helper property to read it from the current game.

File: src/microbe_stage/MicrobeStage.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeStage.cs:313]: Doesn't this print a ton of info? I think maybe always logging this isn't the best to keep log clutter down a bit...

File: src/microbe_stage/MicrobeStage.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeStage.cs:358]: This should just create a new world settings as this is called (I think) only when playing the scene directly from Godot.

File: src/microbe_stage/MicrobeStage.cs
Comment (Oliveriver) [reply src/microbe_stage/MicrobeStage.cs:313]: I wouldn't say it logs a lot, only three lines.

image

I definitely think this needs to be logged somewhere, as otherwise we wouldn't have a record of the player's world settings if we just receive their logs. But if you'd prefer it to be logged somewhere different than main output, I can change that.

File: src/microbe_stage/ProcessSystem.cs
Comment (hhyyrylainen) [inline src/microbe_stage/ProcessSystem.cs:142]: Would it be possible to fix the code structure so that the process system is not given invalid data? I'd much prefer a "proper" design where this workaround here could be avoided.

File: src/microbe_stage/MicrobeStage.cs
Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeStage.cs:313]: Well we don't log the players overall settings either... But fine, let's have the settings in the output for now, I can always remove the registry load prints to reduce the amount of logs a debugging session prints.

File: src/microbe_stage/editor/CellEditorComponent.GUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/CellEditorComponent.GUI.cs:293]: This kind of loop isn't very uniform with style elsewhere.
Could you reformat this and the next few loops to be like:
```
foreach (var entry in allPartSelectionElements)
...
```
or `foreach (var item in allPartSelectionElements.Items)`?

File: src/microbe_stage/editor/CellEditorComponent.GUI.cs
Comment (hhyyrylainen) [inline src/microbe_stage/editor/CellEditorComponent.GUI.cs:298]: What's the difference to `UpdateTooltipMPCostFactors` seems like there's now two methods that basically do the same thing... I think these should be combined to reduce (my) confusion.

File: src/general/NewGameSettings.cs
Comment (hhyyrylainen) [inline src/general/NewGameSettings.cs:116]: This is a bit long one long block of variables. Usually our scenes have "logical" grouping (for example based on the subtab) for variables like this to make it easier to read and find related variables.

File: src/microbe_stage/MicrobeStage.cs
Comment (Oliveriver) [reply src/microbe_stage/MicrobeStage.cs:123]: The issue then is adding the world settings to the MicrobeStage object, which I did in NewGameSettings.OnConfirmPressed(). As it stands, trying to set the WorldSettings variable from GameWorld in the same place throws an error since CurrentGame is null at that point.

File: src/general/NewGameSettings.cs
Comment (hhyyrylainen) [inline src/general/NewGameSettings.cs:231]: I think a safety OrderBy(p => p.Index) should be added. Or does the simulation parameters already sort and ensure no duplicate index values exist?

File: src/general/NewGameSettings.cs
Comment (hhyyrylainen) [inline src/general/NewGameSettings.cs:490]:
```suggestion
    private void UpdateSelectedDifficultyPresetControl()
```
I was a bit confused what this method did at once, so I suggest a bit more descriptive name.

File: src/general/NewGameSettings.cs
Comment (hhyyrylainen) [inline src/general/NewGameSettings.cs:614]: I think the default case should print an error (but then return procedural), so I'd like to see that and `case 0` added.

File: src/general/NewGameSettings.cs
Comment (hhyyrylainen) [inline src/general/NewGameSettings.cs:647]: Can't this and OnGameSeedChangedFromAdvanced be combined? Same as most of the other callbacks or is there a reason why this is this way?

File: src/general/NewGameSettings.cs
Comment (hhyyrylainen) [inline src/general/NewGameSettings.cs:243]: Why doesn't SetSeed update gameSeed and gameSeedAdvanced automatically (assuming the given new seed is valid)?

**Comment by hhyyrylainen** [comment]
I've now finished reviewing all the changes in this PR, so those were all of my findings.

File: src/microbe_stage/MicrobeStage.cs
Comment (Oliveriver) [reply src/microbe_stage/MicrobeStage.cs]: CurrentGame is not initialised at the time we exit the game settings screen. It is only added to MicrobeStage in SetupStage(), which is called after we switch to the scene. So the key problem is that we can't just pass the world settings to CurrentGame at this point.

I tried passing the settings to the new game _after_ switching the scene, but then many things (starting patch, cloud densities, etc.) initialised without the correct world settings.

I agree that GameWorld should be the source of truth for the world settings, but I can't figure out a way of initialising everything correctly without passing a WGS object to MicrobeStage itself. So I've changed this variable to InitialWorldSettings, used only for this initial setup. Everything else should use the WorldSettings object in GameWorld.

File: src/general/NewGameSettings.cs
Comment (Oliveriver) [reply src/general/NewGameSettings.cs]: Ah, after making this change I've realised why I couldn't do that. Since a combined method sets the text of both text inputs, it would in particular overwrite the text the user has just entered. Although it overwrites to the correct value, it places the caret back at the start of the box and won't let you change the caret position. So unless you have a suggestion for that, I'll revert to the way it was.

File: src/microbe_stage/MicrobeStage.cs
Comment (Oliveriver) [reply src/microbe_stage/MicrobeStage.cs]: I discovered while testing InitialWorldSettings that we do need to pass the settings object here, even in the normal game mode. So I've kept it.

Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeStage.cs]: I think this is related to: https://github.com/Revolutionary-Games/Thrive/pull/3389#discussion_r902677928

File: src/microbe_stage/MicrobeStage.cs
Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeStage.cs]: I couldn't help myself so I went ahead and programmed and committed this kind of change. Also added 2 more words to the dictionary to suppress some warnings.

**Comment by hhyyrylainen** [comment]
I just playtested this for the first time, and I really think the "basic view" and "advanced view" should just be named "basic" and "advanced", they are clear enough from the context and I think adding the "view" to the end sounds worse.

There is some kind of weird flickering going on when hovering with the mouse on the random seed where the cursor flickers between the caret and normal. I've not noticed this happening with any other text boxes in the game.

Also made another commit to fix the place that allowed null world settings and based on that I found that some parts of auto-evo didn't take the world settings into account, so I fixed those places as well.

Besides that "basic view" and the things I fixed myself, I didn't notice problems.

**Comment by Oliveriver** [comment]
Alright, thanks for that. I was working on those changes myself and had pretty much exactly the same changes to all the Auto-Evo code, but I'll defer to your commit.

I can't reproduce the flickering issue. Do you think it's a big enough problem to investigate further?

**Comment by hhyyrylainen** [comment]
The flickering might be platform specific (linux only) or might only happen in windowed mode (if you are playing in fullscreen). If it can't be solved, I'll open an issue for fixing it later.

**Comment by Oliveriver** [comment]
I've been playing in fullscreen, but having just switched to windowed mode, I still can't see an issue. So maybe it is Linux-specific.

**Review by hhyyrylainen — APPROVED** [review]
As far as I recall, this is now fully done. I'll merge this tomorrow morning unless someone has anything still to say.

**Pull request merged**
